c/src/lib/libbsp/arm/shared/stm32fxxxx/spi/hal-spi.c

```diff
@@ -0,0 +1,507 @@
+/**
+ * @file hal-spi.c
+ *
+ * @ingroup spi
+ *
+ * @brief SPI driver for the STM32xxxx series processors. Provides a
+ * SPI driver that registers with cpukit/dev/spi as a node in the
+ * filesystem.
+ *
+ */
+
+/*
+ * Copyright (c) 2015 Vecna Technologies, Inc.
+ *
+ * The license and distribution terms for this file may be
+ * found in the file LICENSE in this distribution or at
+ * http://www.rtems.com/license/LICENSE.
+ */
+
+//================== STMF32 Support Functions =================================
+#include <rtems.h>
+#include <hal-utils.h>
+#include <stm32f-processor-specific.h>
+#include <bspopts.h>
+
+#include stm_processor_header(TARGET_STM_PROCESSOR_PREFIX)
+#include stm_header(TARGET_STM_PROCESSOR_PREFIX, dma)
+#include stm_header(TARGET_STM_PROCESSOR_PREFIX, spi)
+#include stm_header(TARGET_STM_PROCESSOR_PREFIX, rcc)
+#include stm_header(TARGET_STM_PROCESSOR_PREFIX, gpio)
+
+#include <hal-spi.h>
+#include <dev/spi/spi.h>
+
+#include <math.h>
+#include <rtems/irq-extension.h>
+
+/**
+ * SPI Instance enabled / disabled
+ */
+bool SPI_Enable[MAX_SPI_INSTANCES] =
+{
+#ifndef STM32_ENABLE_SPI1
+      false,
+#else
+      STM32_ENABLE_SPI1,
+#endif
+#ifndef STM32_ENABLE_SPI2
+      false,
+#else
+      STM32_ENABLE_SPI2,
+#endif
+#ifndef STM32_ENABLE_SPI3
+      false,
+#else
+      STM32_ENABLE_SPI3,
+#endif
+#ifndef STM32_ENABLE_SPI4
+      false,
+#else
+      STM32_ENABLE_SPI4,
+#endif
+#ifndef STM32_ENABLE_SPI5
+      false,
+#else
+      STM32_ENABLE_SPI5,
+#endif
+#ifndef STM32_ENABLE_SPI6
+      false,
+#else
+      STM32_ENABLE_SPI6,
+#endif
+};
+
+/**
+ * SPI pin configuration for each instances.
+ * Pin and port details to be populated in configure.ac file
+ */
```

Comment:
Silly question, but are we so tight on memory that we need to suppress the structs in the array for this info? It seems like to me you can just have the array with all of your initialization information and then check against the enables wherever the initialization happen. I think there are ways to do this with less ifdefs that don't take up much more memory too.